From d24c536ba502d65e6dcc243778bf70e93afff3b3 Mon Sep 17 00:00:00 2001
From: Julien Thomas <jthomas@zenetys.com>
Date: Thu, 18 Jul 2024 19:52:50 +0200
Subject: [PATCH] ossl: Fix #pragma GCC diagnostic not allowed inside functions

Some old compilers, eg. GCC 4.4.7 on el6 do not support #pragma GCC
diagnostic in functions. Rsyslog has PRAGMA_* macros to handle those
cases, so let's use them.
---
 runtime/net_ossl.c | 18 +++++++++---------
 runtime/rsyslog.h  |  8 ++++++--
 2 files changed, 15 insertions(+), 11 deletions(-)

diff --git a/runtime/net_ossl.c b/runtime/net_ossl.c
index 77d21414e..987d1d520 100644
--- a/runtime/net_ossl.c
+++ b/runtime/net_ossl.c
@@ -220,8 +220,8 @@ osslGlblInit(void)
 	ERR_load_crypto_strings();
 #endif
 
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+PRAGMA_DIAGNOSTIC_PUSH
+PRAGMA_IGNORE_Wdeprecated_declarations
 
 	// Initialize OpenSSL engine library
 	ENGINE_load_builtin_engines();
@@ -243,7 +243,7 @@ osslGlblInit(void)
 	}
 	// Free the engine reference when done
 	ENGINE_free(osslEngine);
-#pragma GCC diagnostic pop
+PRAGMA_DIAGNOSTIC_POP
 }
 
 /* globally de-initialize OpenSSL */
@@ -1094,8 +1094,8 @@ RSYSLOG_BIO_debug_callback(BIO *bio, int cmd, const char __attribute__((unused))
 
 #if OPENSSL_VERSION_NUMBER >= 0x10100000L
 // Requires at least OpenSSL v1.1.1
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wunused-parameter"
+PRAGMA_DIAGNOSTIC_PUSH
+PRAGMA_IGNORE_Wunused_parameter
 static int
 net_ossl_generate_cookie(SSL *ssl, unsigned char *cookie, unsigned int *cookie_len)
 {
@@ -1118,7 +1118,7 @@ net_ossl_generate_cookie(SSL *ssl, unsigned char *cookie, unsigned int *cookie_l
 
 	return 1;
 }
-#pragma GCC diagnostic pop
+PRAGMA_DIAGNOSTIC_POP
 
 static int
 net_ossl_verify_cookie(SSL *ssl, const unsigned char *cookie, unsigned int cookie_len)
@@ -1149,8 +1149,8 @@ net_ossl_init_engine(__attribute__((unused)) net_ossl_t *pThis)
 	const char *engine_id = NULL;
 	const char *engine_name = NULL;
 
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+PRAGMA_DIAGNOSTIC_PUSH
+PRAGMA_IGNORE_Wdeprecated_declarations
 	// Get the default RSA engine
 	ENGINE *default_engine = ENGINE_get_default_RSA();
 	if (default_engine) {
@@ -1188,7 +1188,7 @@ net_ossl_init_engine(__attribute__((unused)) net_ossl_t *pThis)
 	} else {
 		DBGPRINTF("net_ossl_init_engine: use openssl default Engine");
 	}
-#pragma GCC diagnostic pop
+PRAGMA_DIAGNOSTIC_POP
 
 	RETiRet;
 }
diff --git a/runtime/rsyslog.h b/runtime/rsyslog.h
index 6446b13ff..970df76ae 100644
--- a/runtime/rsyslog.h
+++ b/runtime/rsyslog.h
@@ -108,14 +108,18 @@
 						_Pragma("GCC diagnostic ignored \"-Wunknown-attribute\"")
 	#define PRAGMA_IGNORE_Wformat_nonliteral \
 						_Pragma("GCC diagnostic ignored \"-Wformat-nonliteral\"")
-	#define PRAGMA_IGNORE_Wdeprecated_declarations \
-						_Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
 	#if  __GNUC__ >= 5
+		#define PRAGMA_IGNORE_Wdeprecated_declarations \
+			_Pragma("GCC diagnostic ignored \"-Wdeprecated-declarations\"")
+		#define PRAGMA_IGNORE_Wunused_parameter \
+			_Pragma("GCC diagnostic ignored \"-Wunused-parameter\"")
 		#define PRAGMA_DIAGNOSTIC_PUSH \
 			_Pragma("GCC diagnostic push")
 		#define PRAGMA_DIAGNOSTIC_POP \
 			_Pragma("GCC diagnostic pop")
 	#else
+		#define PRAGMA_IGNORE_Wdeprecated_declarations
+		#define PRAGMA_IGNORE_Wunused_parameter
 		#define PRAGMA_DIAGNOSTIC_PUSH
 		#define PRAGMA_DIAGNOSTIC_POP
 	#endif
-- 
2.21.1

